sv.c

```diff
@@ -9635,15 +9906,75 @@ C<L</sv_setsv_flags>>.
 SV *
 Perl_sv_mortalcopy_flags(pTHX_ SV *const oldstr, U32 flags)
 {
-    SV *sv;
+    SV * old = (oldstr) ? oldstr : &PL_sv_undef;
+    SV *dsv;
+    const svtype old_type = SvTYPE(old);
 
-    if (flags & SV_GMAGIC)
-        SvGETMAGIC(oldstr); /* before new_SV, in case it dies */
-    new_SV(sv);
-    sv_setsv_flags(sv,oldstr,flags & ~SV_GMAGIC);
-    PUSH_EXTEND_MORTAL__SV_C(sv);
-    SvTEMP_on(sv);
-    return sv;
+#if NVSIZE <= IVSIZE
+    if (old_type <= SVt_NV) {
+#else
+    if (old_type <= SVt_IV) {
+#endif
+        new_SV(dsv);
+        U32 sflags = SvFLAGS(old);
+#if NVSIZE <= IVSIZE
+        switch(sflags & (SVf_IOK|SVf_ROK|SVf_NOK)) {
+#else
+        switch(sflags & (SVf_IOK|SVf_ROK)) {
+#endif
+            case SVf_ROK:
+                SET_SVANY_FOR_BODYLESS_IV(dsv);
+                SvFLAGS(dsv) = SVt_IV|SVf_ROK;
+                dsv->sv_u.svu_rv = SvREFCNT_inc(SvRV(old));
+                break;
+            case SVf_IOK:
+                SET_SVANY_FOR_BODYLESS_IV(dsv);
+                /* both src and dst are <= SVt_IV, so sv_any points to the
+                 * head; so access the head directly
+                 */
+                assert(    &(old->sv_u.svu_iv)
+                        == &(((XPVIV*) SvANY(old))->xiv_iv));
+                assert(    &(dsv->sv_u.svu_iv)
+                        == &(((XPVIV*) SvANY(dsv))->xiv_iv));
+                dsv->sv_u.svu_iv = old->sv_u.svu_iv;
+                SvFLAGS(dsv) = SVt_IV|(sflags & (SVf_IOK|SVp_IOK|SVf_IVisUV));
+                break;
+#if NVSIZE <= IVSIZE
+            case SVf_NOK:
+                SET_SVANY_FOR_BODYLESS_NV(dsv);
+                assert(    &(old->sv_u.svu_nv)
+                    == &(((XPVNV*) SvANY(old))->xnv_u.xnv_nv));
+                assert(    &(dsv->sv_u.svu_nv)
+                    == &(((XPVNV*) SvANY(dsv))->xnv_u.xnv_nv));
+                dsv->sv_u.svu_nv = old->sv_u.svu_nv;
+                SvFLAGS(dsv) = SVt_NV|SVf_NOK|SVp_NOK;
+                break;
+#endif
+        }
+    } else if (old_type == SVt_PV) {
+        dsv = newSV_type(SVt_PV);
+        if (SvFLAGS(old) & SVp_POK) {
+            assert(!SvIsCOW_static(old));
+            S_sv_freshcopy_POK(aTHX_ dsv, old, flags);
+        } else if (SvFLAGS(old) & SVf_ROK) {
+            SvFLAGS(dsv) |= SVf_ROK;
+            SvRV_set(dsv, SvREFCNT_inc(SvRV(old)));
+        }
```

Comment:
weak ref applies here or not?

Reply:
A copy of a weak ref isn't weak itself, I don't think there needs to be any special handling of weak references here.

Reply:
If so, just ignore anything I said about weak refs. That topic is N/A for this PR and what this PR intentions are. This PR isn't supposed to have any visible PP or CPAN XS behavior changes. This PR's only purpose is cutting out turing machine logic steps and cutting out `-O2` machine code steps, that no C compiler can ever optimize away. Only the human end user can cut them out. I dont think this PR should be trying to touch or do anything new, in the situation of, If official master branch blead perl isnt doing "XYZ behavior" regarding overload.pm/tie/MG* right now, then this PR is going introduce "XYZ" either. "XYZ" topic is for another future PR/ticket/thread.